### ISSUE TITLE
(symbolic/soft link) shared resourcepaks / shaderpacks / other folders

### ISSUE THREAD

**Issue by NoelGisler** [PrismLauncher/PrismLauncher#38]
### Role

i play multiple instances on Primslauncher

### Suggestion

i want the ability to share folders (mainly resourcepacks and shaderpacks) over multiple instances

### Benefit

i won't always have to download / copy all the packs into the new instance (and probably save a few MB)

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

i imagine it somewhat like this:

./instances/modded-instance-1/resourcepacks gets softlinked to ./shared-folders/shared-group-1/shared-resourcepacks

i already did that on my original PolyMC install, although i didn't go with groups there yet, but instead just one folder for all instances to share. it worked really well in that environment (tested on Fedora Linux and NixOS).

the advantages of groups is, that those who want can make a "all instances group" and the rest can make their own groups, if they need them for some reason. maybe some modpacks need special texture packs you don't want to clog up everything, maybe some other reason.

it could possibly also be used to share servers or even worlds.

**Comment by JustLetterV** [comment]
There is mod that can do what you are asking for:
https://modrinth.com/mod/shared-resources

**Comment by pseudoforceyt** [comment]
What if some (or a lot) instances you are using are not fabric or forge?
I would like to see this feature being implemented (if possible)

**Comment by Octelly** [comment]
> What if some (or a lot) instances you are using are not fabric or forge? I would like to see this feature being implemented (if possible)

yeah and the mod is 1.16+ if I'm reading it correctly

**Comment by JustLetterV** [comment]
I agree, this feature would be really nice to have. I just shared this mod just in case OP needs it :) Also it is only 1.16+ sadly.

**Comment by yavko** [comment]
This could be cool to sym link mods from a central directory to save tons of space, which that mod also doesn't support.

**Comment by Continous** [comment]
It also makes far more sense to do this from the launcher, as it allows the instance/client to be naive to the folder structure, or the actual location of the files. It also doesn't require a mod. It'd also be useful if something like server instances were added.

**Comment by Un1q32** [comment]
> There is mod that can do what you are asking for: https://modrinth.com/mod/shared-resources

this should be a feature of the launcher, not a mod

**Comment by Inconn** [comment]
I also want this feature built into the launcher since I have some large resource packs that I really don't want to create copies of but I don't want to have to manage the symlinks myself. It also would be cool to have a central mod folder where mods get symlinked into the mods folder to save space if you reuse mods but I think that only having a global resources and shaders packs folder would be good enough since it wouldn't make as much of a difference (for me, at least). Though I don't know how you'd do it on Windows, i'm personally a Linux user.

**Comment by Continous** [comment]
> Though I don't know how you'd do it on Windows, i'm personally a Linux user.

Symbollic and hardlinks exist on both platforms, and as relevant on most filesystems. Implementing a simple check to see if symbolic links are available can be implemented. The biggest hurdle on Windows is that placing symbolic/hardlinks required administrative access last I used it.

**Comment by yavko** [comment]
> > Though I don't know how you'd do it on Windows, i'm personally a Linux user.
> 
> Symbollic and hardlinks exist on both platforms, and as relevant on most filesystems. Implementing a simple check to see if symbolic links are available can be implemented. The biggest hurdle on Windows is that placing symbolic/hardlinks required administrative access last I used it.

You don't need admin for it afaik
EDIT: I think there is even a powershell command for it, but I dont use windows so I can't confirm

**Comment by Inconn** [comment]
From what i've searched you need admin for it on Windows, might test it later on another pc

**Comment by yavko** [comment]
Okay from deeper research it is a group permissions that is required for it `SeCreateSymbolicLinkPrivilege`, proof windows is a legacy OS 👏🏻 
EDIT: Actually its been fixed in newer windows versions, IF you have developer mode enabled, but I don't think thats the case anymore either

**Comment by Continous** [comment]
Yeah, regardless, since some, albeit niche, filesystems and conditions may make symbolic or hardlinks break, there should be a check if such a feature is implemented.

**Comment by Inconn** [comment]
Perhaps it's possible to tell minecraft to look elsewhere or to look in multiple places for resource packs? I know a mod like that exists.

**Comment by Continous** [comment]
Not without a mod or runtime hook. Both which can have adverse effects, and/or be otherwise undesirable. Symlinks/hardlinks are a far more agnostic solution, and avoids any reliance on Minecraft having the capacity. If, for example, certain mods rely on local resources stored in the main minecraft folder, haphazardly redirecting things with a mod may cause issues. A symlink, would allow real/original files to coexist without special consideration.

**Comment by Inconn** [comment]
Yeah, links would be the best choice but considering that most people use Windows there won't be a lot of people to benefit from this feature unless they changed their permissions, which most people probably wouldn't want to do, or ran as administrator, which is obviously not a good idea. I still want this feature but the average person wouldn't be able to use it.

**Comment by yavko** [comment]
It isn't a admin only thing anymore, but developer mode has to be enabled as it is built as a new api for non admin usage to be used in dev tools, and that's a one time thing. The best way is to show a popup that this feature requires dev mode if not available on windows, but imo prism launcher should keep priority for modern OSes.

**Comment by Ryex** [comment]
we may be able to take a multiple path method of achieving this.

the ability to symlink files would also be useful for de-duplicating mod files and conserving disk space.

Apple, Linux and other *nix will let us create symbolic links easily in-process

On windows this requires admin access, or did until very recently and even then requires group policy changes; full support of our targeted versions would require admin access.
on the windows OSes we could then delegate to a separate `PrismLinkFileDaemon.exe` which is run with elevated privilege via `ShellExacute` and the `runas` verb. This would require the user accept the UAC prompt and communication with this daemon would have to be via named pipes but it is in theory possible.

if the filesystem in question does not allow symbolic link (ie FAT) it could fall back to a file copy.

**Comment by ColonelGerdauf** [comment]
An alternative that I can see is some way to have the Minecraft instances look in the shared folder in the first place.

This may sound far fetched, but this is how it worked with the Mojang launcher, at least with shaders.

**Comment by Continous** [comment]
> An alternative that I can see is some way to have the Minecraft instances look in the shared folder in the first place.
> 
> This may sound far fetched, but this is how it worked with the Mojang launcher, at least with shaders.

There's no way to do this without undertaking significant work to create a runtime modifying program, or some virtual filesystem, at which point just doing hard/soft links is better.

**Comment by ColonelGerdauf** [comment]
I am honestly torn on resource packs, as version distinctions can break the feel of the game, unless there is a system that can automatically filter the correct versions, which is a lot more challenging than it sounds. This might work with the legacy texture packs, but the risk is still there.

But it makes me wonder how the default launcher made it work with a singular 'shaderpack' folder at the root of the userdata.

**Comment by Continous** [comment]
> I am honestly torn on resource packs, as version distinctions can break the feel of the game, unless there is a system that can automatically filter the correct versions, which is a lot more challenging than it sounds. This might work with the legacy texture packs, but the risk is still there.
> 
> But it makes me wonder how the default launcher made it work with a singular 'shaderpack' folder at the root of the userdata.

So a few things here,

First; the only major distinction between older and newer versions is where textures are placed, and maybe the texture names. This is something already ironed out/squared away within minecraft as older pre-atlased texture packs have a different folder, Minecraft automatically informs a user of incorrect resource pack versions, and most mods due the same. As for visual differences, that's irrelevant. Users are expected to manage resource packs on their own, if they choose to use a resource/texture pack from the wrong version that's their perogative.

Second, the default launcher doesn't do anything special with the shaderpack folder. That's Minecraft itself having different file locations hardcoded in for that specific folder.

**Comment by Continous** [comment]
Sorry to double up, but given that Prism already distinguished Minecraft versions, it should be easy to determine which folders ought to be considered and which ought not to be. Similarly, there could be simple methods to add/remove individual packs/files instead of slapping whole folders.

**Comment by superon** [comment]
i'm confused, is this in the pipeline or not atm?  
also, as i was redirected here by having my suggestion closed as being a duplicate of this i'd like to stress that i was specifically writing about having a global mods folder (to have prism read from) creating a "link" to a mod within via the prism gui.   i've much less interest in the user creating symlinks per whatever process their OS provides

**Comment by Trial97** [comment]
Yes, this is planned (when it is implemented)
And it is still the same issue; like I see no point tracking it as a separate issue when, at the end of the day, they solve the same problem. 

PS. I do not know why it is so hard to create it yourself manually if you need it so badly.

**Comment by Continous** [comment]
It is not hard to do manually. But the entire point of a mod pack manager is to do the tedious stuff for you. It's not really polite to dismiss people's issues like that especially after acknowledging it is seen as an active and real issue.

**Comment by superon** [comment]
not like i need it badly, it would be super useful in the longrun though.   i've looked into symlinks for about the past hour and don't consider that a very useful option for most users :/

**Comment by Trial97** [comment]
Sorry, I did not want to dismiss your issue. I'm just trying to group issues that are similar enough to keep a cleaner issue tracker and see how many people wish for that specific feature.

**Comment by Continous** [comment]
> Sorry, I did not want to dismiss your issue. I'm just trying to group issues that are similar enough to keep a cleaner issue tracker and see how many people wish for that specific feature.

Hey, I can appreciate that. And everyone's situation is unique. Sometimes we gotta course correct when we get a bit too harsh with our language.

**Comment by IntangibleMatter** [comment]
How is this still open after almost three years? It feels like it should be easy to have a "central" resource pack folder shared across instances.

**Comment by Continous** [comment]
> How is this still open after almost three years? It feels like it should be easy to have a "central" resource pack folder shared across instances.

Things can often appear deceptively simple or easy.

**Comment by mirk0dex** [comment]
> How is this still open after almost three years? It feels like it should be easy to have a "central" resource pack folder shared across instances.

then you implement it champ.

**Comment by IntangibleMatter** [comment]
> then you implement it champ

I would but I'm a games programmer, filesystem stuff is not my area of expertise.

**Comment by yavko** [comment]
> > then you implement it champ
> 
> I would but I'm a games programmer, filesystem stuff is not my area of expertise.

Why not learn then?

**Comment by PmikeGamer** [comment]
> > > then you implement it champ
> > 
> > 
> > I would but I'm a games programmer, filesystem stuff is not my area of expertise.
> 
> Why not learn then?

Maybe they don't have time. Also isn't that the point of github? So other people who know stuff can contribute so not everyone needs to learn everything that's involved?

**Comment by Ryex** [comment]
> > then you implement it champ
> 
> I would but I'm a games programmer, filesystem stuff is not my area of expertise.

This should answer your own question.

it is not, in fact, simple to implement a way to link resource folders across instances in a way that is robust and cross-platform. maintainers and contributors have had higher priorities.

We'll always help review any attempt at the feature and help it get approved.

**Comment by BeeFox-sys** [comment]
yeah i think the big thing to my understanding is that symlinks are different between OSes, and off the top of my head there aren't any easy ways to do it that's not "check the OS and do it differently for each", i would really like this one day (and i have noticed that over time we have gotten *some* support for it! namely in advanced copy options), but its definitely not an easy thing so if you would want to try learning it and give it a shot im sure the maintatiners would be more then happy to look at the code c:

(and i know saying that is kinda hypocritical because im not doing that myself, but i have my own situations etc)

**Comment by InvitedToHell** [comment]
Would love to see this too!

**Comment by ClairDePrince** [comment]
Do we have any news about this?

**Comment by InvitedToHell** [comment]
Not that I know of any :(

**Comment by MidyGamy** [comment]
I'd suggest if it's done to also allow making custom folders for if a new mod adds a folder (instead of just the basics ones, example: Figura, CPM and I skip many)
And add a toggle for instances that can isolate an instance from the shared folders
